### ISSUE TITLE
2D Output error

### ISSUE THREAD

**Issue by mwoodson1** [NervanaSystems/neon#208]
I am trying to do something similar to what #145 mentioned with my output y being a 72x72 image. I followed the instructions in that issue to create my ArrayIterator and set my output layer to
`Affine((72,72), init=Gaussian(scale=0.03), bias=None, activation=Rectlin())`

Though when I try to train the network I get the following error:

Traceback (most recent call last):
File "correlation_encoder.py", line 64, in <module>
model.fit(train_set, optimizer=opt, num_epochs=num_epochs, cost=cost, callbacks=callbacks)
  File "/home/mwoodson/installs/neon/neon/models/model.py", line 141, in fit
    self.initialize(dataset, cost)
  File "/home/mwoodson/installs/neon/neon/models/model.py", line 103, in initialize
    cost.initialize(prev_input)
  File "/home/mwoodson/installs/neon/neon/layers/layer.py", line 1336, in initialize
    parallelism=self.prev_layer.parallelism)
  File "/home/mwoodson/installs/neon/neon/backends/backend.py", line 536, in iobuf
    persist_values=persist_values)
  File "/home/mwoodson/installs/neon/neon/backends/nervanagpu.py", line 1192, in zeros
    rounding=self.round_mode)._assign(0)
  File "/home/mwoodson/installs/neon/neon/backends/nervanagpu.py", line 135, in __init__
    self.gpudata = allocator(self.nbytes)
Boost.Python.ArgumentError: Python argument types in
    pycuda._driver.mem_alloc(tuple)
did not match C++ signature:
    mem_alloc(unsigned long)

Help here would be much appreciated. For reference my input dim is (128,15552) and output dim is (128,72,72).

**Comment by apark263** [comment]
The number of outputs for affine should be a number rather than a tuple. In
this case you actually want 72x72x3, since it seems like your input is a 3
channel image, and you need to match your input size for a SSE cost

That should get rid of the error, but im not sure itll get the
reconstruction performance you want

**Comment by mwoodson1** [comment]
Thanks for the quick response. I am not exactly creating an autoencoder so reconstruction is not the objective. Since would like the output to simply be a 72x72 grayscale image, changing my code to 
`Affine(72*72, init=Gaussian(scale=0.03), bias=None, activation=Rectlin())`
would mean I would have to change my ArrayIterator initialization correct?

For reference here is what it is now:
`train_set = ArrayIterator(X_train, y_train, nclass=y_train.shape[1], make_onehot=False, lshape=(3,72,72))`
which was taken from #145 .

**Comment by apark263** [comment]
Yes, easiest thing to do would actually be to do some operation on the data
itself (X_train) to make it single channel. There is specific weighting of
the color channels that needs to be done to convert from RGB into
grayscale. Once that is done change the shape tuple in ArrayIterayor to be
(1,72,72)

Note that the shape is only important if you are feeding the data into a
layer that needs some notion of shape, e.g. Pooling or convolution. If the
data is going straight into another affine layer, the shape is ignored

**Comment by mwoodson1** [comment]
Ah sorry I think I should clarify. I would still like the input to be RGB but my output would be grayscale(my output is not neccesarily an image but a 2D array that can be considered an image). The idea is I would like to learn a single filter but from what I have seen neon does not allow a Conv layer with only 1 filter channel. Thus I changed it to an Affine layer with size (72,72) as that was the filter dimension I desired.

I have now changed the Affine layer to have `nout=72*72`  and my `nclass=72*72` in my ArrayIterator. Reshaping the output of the Affine layer to 72x72 will be exactly what I desire correct?
